default-constructed-buffers/default-constructed-buffers.md

```diff
@@ -0,0 +1,88 @@
+# Default-constructed Buffers
+
+Proposal ID      | TBC
+---------------- | ---------------------------
+Name             | Default-constructed Buffers
+Date of Creation | 2019-08-20
+Target           |
+Status           | WIP
+Author           | Duncan McBain [Duncan McBain](mailto:duncan@codeplay.com)
+Contributors     |
+
+## Description
+
+In SYCL, there are no buffer constructors that take no arguments, which means
+that `cl::sycl::buffer` cannot be default-constructed. However, this limits the
+use of buffers in some interfaces. Consider a library function that has an
+optional user-provided allocation for temporary data. The library can avoid
+allocations if the library user so desires by providing a buffer here, but how
+should the user indicate that allocation is allowed? A natural way would be to
+provide a default argument in the API function:
+
+```c++
+template <typename Allocation_t>
+void foo(Allocation_t input_a,
+         Allocation_t input_b,
+         Allocation_t temp = Allocation_t{});
+```
+
+For `Allocation_t` types that are pointer-like, the library can check for
+`nullptr` and allocate if this condition is true. A similar mechanism that
+allows `cl::sycl::buffer` to be constructed and checked for usability would
+aid their use in this style of generic interface.
+
+## Proposal
+
+The `cl::sycl::buffer` class should be augmented with an additional constructor
+that takes no arguments, which initialises the buffer with a zero-size range.
+```c++
+namespace cl {
+namespace sycl {
+template <typename T, int dimensions = 1,
+typename AllocatorT = cl::sycl::buffer_allocator>
+class buffer {
+  buffer();
+
+  bool is_valid() const noexcept;
+
+  explicit operator bool() const noexcept;
+};
+} // namespace sycl
+} // namespace cl
+```
+The template arguments should remain the same, so that the argument can be
+rebound to a new `buffer` instance later using the copy constructor.
+
+The `is_valid()` call would allow the programmer to query whether or not
```

Comment:
We should also describe what should happen if you try to use an invalid buffer, i.e. call `get_access`, this should probably throw an exception.